### PR TITLE
Remove redundancy in vanilla/zsa circuits

### PR DESCRIPTION
**Pull request by ConstanceBeguier**
This PR contains the following modifications:
- Unified `NoteCommitConfig` and `Config` to have the same configs for both Vanilla and ZSA circuits
- Numerous functions and structures initially defined separately for Vanilla and ZSA have been unified into a single definition, incorporating the generic parameter `Lookup`.

### CODE REVIEW

File: src/circuit/circuit_vanilla/derive_nullifier.rs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+            add_chip,
+            ecc_chip,
+            rho,
+            psi,
+            cm,
+            nk,
+        )
+        .map(|res| res.extract_p())
+    }
+}
```
Comment (PaulLaux) [inline src/circuit/circuit_vanilla/derive_nullifier.rs:54]: This file adds very little.

File: src/circuit/circuit_zsa/derive_nullifier.rs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+            cond_swap_chip.mux_on_points(
+                layouter.namespace(|| "mux on nf"),
+                &split_flag,
+                nf.inner(),
+                split_note_nf.inner(),
+            )?,
+        )
+        .extract_p())
+    }
+}
```
Comment (PaulLaux) [inline src/circuit/circuit_zsa/derive_nullifier.rs:78]: I would move all logic from this file and `circuit_vanilla/derive_nullifier.rs` to `circuit/derive_nullifier.rs`.
lets have optional params: `cond_swap_chip` and `split_flag`. The extra functionality can be decided for `None` or `Some`
something like
```rust
    struct zsa_nullifier_params{
        pub cond_swap_chip: CondSwapChip<pallas::Base>,
        pub split_flag: AssignedCell<pallas::Base, pallas::Base>,
    }
   pub(in crate::circuit) fn derive_nullifier<
    PoseidonChip: PoseidonSpongeInstructions<pallas::Base, poseidon::P128Pow5T3, ConstantLength<2>, 3, 2>,
    AddChip: AddInstruction<pallas::Base>,
    EccChip: EccInstructions<
        pallas::Affine,
        FixedPoints = OrchardFixedBases,
        Point = EccPoint,
        Var = AssignedCell<pallas::Base, pallas::Base>,
    >,
>(
    layouter: &mut impl Layouter<pallas::Base>,
    poseidon_chip: PoseidonChip,
    add_chip: AddChip,
    ecc_chip: EccChip,
    rho: AssignedCell<pallas::Base, pallas::Base>,
    psi: &AssignedCell<pallas::Base, pallas::Base>,
    cm: &Point<pallas::Affine, EccChip>,
    nk: AssignedCell<pallas::Base, pallas::Base>,
    zsa_params: Option<zsa_nullifier_params>,
  ) -> Result<X<pallas::Affine, EccChip>, plonk::Error> {
```

File: src/circuit/circuit_vanilla/gadget.rs [hunk 3]
```diff
@@ -1,190 +1,6 @@
...
-    // Add cm to multiplied fixed base to get nf
-    // cm + [poseidon_output + psi] NullifierK
-    cm.add(layouter.namespace(|| "nf"), &product)
-        .map(|res| res.extract_p())
-}
-
 pub(in crate::circuit) use super::commit_ivk::gadgets::commit_ivk;
+pub(in crate::circuit) use super::derive_nullifier::gadgets::derive_nullifier;
 pub(in crate::circuit) use super::note_commit::gadgets::note_commit;
+pub(in crate::circuit) use super::value_commit_orchard::gadgets::value_commit_orchard;
```
Comment (PaulLaux) [inline src/circuit/circuit_vanilla/gadget.rs:6]: can we remove this file completely? (x2)

File: src/circuit/circuit_zsa/gadget.rs [hunk 4]
```diff
@@ -1,245 +1,6 @@
...
-            &split_flag,
-            nf.inner(),
-            split_note_nf.inner(),
-        )?,
-    )
-    .extract_p())
-}
-
 pub(in crate::circuit) use super::commit_ivk::gadgets::commit_ivk;
+pub(in crate::circuit) use super::derive_nullifier::gadgets::derive_nullifier;
```
Comment (PaulLaux) [inline src/circuit/circuit_zsa/gadget.rs:4]: can we remove this file?

**Review by PaulLaux — COMMENTED** [review]
Added minor comments and suggestions

File: src/circuit/circuit_vanilla/derive_nullifier.rs (hunk 1 above)
Comment (ConstanceBeguier) [reply src/circuit/circuit_vanilla/derive_nullifier.rs]: I removed circuit_vanilla/derive_nullifier.rs and circuit_zsa/derive_nullifier.rs

File: src/circuit/circuit_zsa/derive_nullifier.rs (hunk 2 above)
Comment (ConstanceBeguier) [reply src/circuit/circuit_zsa/derive_nullifier.rs]: I moved all logic from circuit_vanilla/derive_nullifier.rs and circuit_zsa/derive_nullifier.rs to circuit/derive_nullifier.rs

File: src/circuit/circuit_vanilla/gadget.rs (hunk 3 above)
Comment (ConstanceBeguier) [reply src/circuit/circuit_vanilla/gadget.rs:6]: I moved all logic from circuit_vanilla/value_commit_orchard.rs and circuit_zsa/value_commit_orchard.rs to circuit/value_commit_orchard.rs

File: src/circuit/circuit_zsa/gadget.rs (hunk 4 above)
Comment (ConstanceBeguier) [reply src/circuit/circuit_zsa/gadget.rs]: Done

**Review by PaulLaux — APPROVED** [review]
Very nice, added one minor comment.

File: src/circuit/derive_nullifier.rs [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+                scalar,
+            )?
+        };
+
+        // Add cm to multiplied fixed base
+        // nf = cm + [poseidon_output + psi] NullifierK
+        let nf = cm.add(layouter.namespace(|| "nf"), &product)?;
+
+        match zsa_params {
+            Some(zsa_params) => {
```
[the remark this answers is not in the document]
Comment (ConstanceBeguier) [reply src/circuit/derive_nullifier.rs]: Done

**Review by PaulLaux — APPROVED** [review]

**Pull request merged**